pages/_app.tsx

```diff
@@ -0,0 +1,93 @@
+/**
+ * @file _app.tsx
+ * Override the main app component.
+ */
+
+import React from 'react';
+import App from 'next/app';
+import Router from 'next/router';
+import UrlSearchParams from 'url-search-params';
+
+class TwApp extends App {
+
+  async componentDidMount() {
+    this.initializeRouteState();
+
+    // Register route change events.
+    Router.events.on('routeChangeComplete', this.handleRouteChangeComplete);
+  }
+
+  componentWillUnmount() {
+    // Remove route changes events.
+    Router.events.off('routeChangeComplete', this.handleRouteChangeComplete);
+  }
+
+  /**
+   * Handle completed route changes.
+   *
+   * Next app router will navigate the browser URL to an app route URL, and has
+   * no way of determining the correct route to use for any given alias URL. We
+   * will provide the alias as a querystring param when rending internal links.
+   * Here we will replace the history state to display the alias in the browser
+   * URL bar, and not the app page route.
+   */
+  handleRouteChangeComplete = url => {
+    const [, urlQuerystring] = url.split('?');
+    const urlParams = new UrlSearchParams(urlQuerystring);
+
+    // If our route contains an alias param, replace current history state
+    // with current state using alias as display URL.
+    if (urlParams.has('alias')) {
+      const alias = urlParams.get('alias');
+      const { router: { asPath, pathname } } = Router;
+
+      window.history.replaceState({
+        url: pathname,
+        as: asPath
+      }, null, alias);
+    }
```

Comment:
Can you explain the need for this feature? I'm assuming it has to do with preserving existing URLs while keeping the routing for this application fairly simple. Is it not possible to look up a story using the existing URL structure (i.e. build an endpoint that can translate those URLs into IDs to fetch from Drupal)?

Reply:
Next's `<Link>` element expects `as` prop values that match the route pattern in the `href` prop. Just using the alias in `as` doesn't work since that URL pattern doesn't match the route, or doesn't have the expected data. Passing the alias via a querystring param was the easiest way I could find to pass the alias into the route in the `<Link>` markup.

There is an API endpoint to look up resource data for an alias, but that only works during server-side rendering of the initial page load. Internal app links manage changes in app state triggered via URL changes using window history API.